fiftyone/utils/multiprocessing.py

```diff
@@ -0,0 +1,376 @@
+class ReduceFcn(object):
+    """Base class for reducers for use with :func:`map_samples`.
+
+    Subclasses may optionally override :meth:`init`, :meth:`update`, and
+    :meth:`finalize` as necessary.
+    """
+
+    def __init__(self, sample_collection, reduce_fcn=None):
+        self.sample_collection = sample_collection
+        self.reduce_fcn = reduce_fcn
+        self.accumulator = None
+
+    def init(self):
+        """Initializes the reducer."""
+        self.accumulator = {}
+
+    def update(self, outputs):
+        """Adds a batch of map outputs to the reducer.
+
+        Args:
+            outputs: a dict mapping sample IDs to map outputs
+        """
+        self.accumulator.update(outputs)
+
+    def finalize(self):
+        """Finalizes the reducer and returns the result, if any.
+
+        Returns:
+            the final output, or None
+        """
+        if self.reduce_fcn is not None:
+            return self.reduce_fcn(self.sample_collection, self.accumulator)
+
+
+def _init_batches(
+    sample_collection,
+    shard_size=None,
+    shard_method="id",
+    num_workers=None,
+):
+    if shard_method == "slice":
+        n = len(sample_collection)
+    else:
+        ids = sample_collection.values("id")
+        n = len(ids)
+
+    if num_workers is None:
+        num_workers = fou.recommend_process_pool_workers()
+
+    # Must cap size of select(ids) stages
+    if shard_method == "id":
+        max_shard_size = fou.recommend_batch_size_for_value(
+            ObjectId(), max_size=100000
+        )
+
+        if shard_size is not None:
+            shard_size = min(shard_size, max_shard_size)
+        elif n > num_workers * max_shard_size:
+            shard_size = max_shard_size
+
+    if shard_size is not None:
+        # Fixed size shards
+        edges = list(range(0, n + 1, shard_size))
+        if edges[-1] < n:
+            edges.append(n)
+    else:
+        # Split collection into exactly `num_workers` shards
+        edges = [int(round(b)) for b in np.linspace(0, n, num_workers + 1)]
+
+    if shard_method == "slice":
+        # Slice batches
+        slices = list(zip(edges[:-1], edges[1:]))
+    else:
+        # ID batches
+        slices = [ids[i:j] for i, j in zip(edges[:-1], edges[1:])]
+
+    num_shards = len(slices)
+    batches = list(
+        zip(
+            range(num_shards),
+            itertools.repeat(num_shards),
+            slices,
+        )
+    )
+
+    num_workers = min(num_workers, num_shards)
+
+    return batches, num_workers
+
+
+def _init_worker(dataset_name, view_stages, m, s, r, p, l):
+    from tqdm.auto import tqdm
+
+    import fiftyone as fo
+    import fiftyone.core.odm.database as food
+    import fiftyone.core.view as fov
+
+    global sample_collection, map_fcn, save, return_outputs, progress
+
+    # Ensure that each process creates its own MongoDB clients
+    # https://pymongo.readthedocs.io/en/stable/faq.html#using-pymongo-with-multiprocessing
+    food._disconnect()
```

Comment:
Haha funny acronym!

Reply:
Do we automatically reconnect the mongodb connection after `map_samples` finishes the computation?

Reply:
This code runs in the worker processes, which are destroyed when the operation completes.